File: src/main/java/dsl/BulkTargetPathBuilder.java
package dsl;

import java.util.Set;
import java.util.function.BiFunction;

/**
 * Builder for specifying the target of bulk mappings.
 */
public class BulkTargetPathBuilder {
    private final TargetBuilder parent;
    private final String sourcePattern;
    private final String targetPath;
    private final Set<String> exclusions;
    private final Set<String> inclusions;
    private final BiFunction<String, Object, Object> transformation;
    private final boolean preserveStructure;
    private final boolean flatten;
    private final String flattenPrefix;

    BulkTargetPathBuilder(TargetBuilder parent, String sourcePattern, String targetPath,
                          Set<String> exclusions, Set<String> inclusions,
                          BiFunction<String, Object, Object> transformation,
                          boolean preserveStructure, boolean flatten, String flattenPrefix) {
        this.parent = parent;
        this.sourcePattern = sourcePattern;
        this.targetPath = targetPath;
        this.exclusions = exclusions;
        this.inclusions = inclusions;
        this.transformation = transformation;
        this.preserveStructure = preserveStructure;
        this.flatten = flatten;
        this.flattenPrefix = flattenPrefix;
    }

    /**
     * Finalize the bulk mapping rule.
     */
    public TargetBuilder end() {
        MappingRule rule;
        if (flatten) {
            rule = new FlattenMappingRule(sourcePattern, targetPath, exclusions,
                    inclusions, transformation, flattenPrefix);
        } else if (preserveStructure) {
            rule = new BulkMappingRule(sourcePattern, targetPath, exclusions,
                    inclusions, transformation, true);
        } else {
            rule = new BulkMappingRule(sourcePattern, targetPath, exclusions,
                    inclusions, transformation, false);
        }
        parent.addRule(rule);
        return parent;
    }
}

File: src/main/kotlin/xyz/mahmoudahmed/dsl/bulk/BulkTargetPathBuilder.kt
package xyz.mahmoudahmed.dsl.bulk

import xyz.mahmoudahmed.dsl.builders.TargetBuilder
import xyz.mahmoudahmed.dsl.core.MappingRule

/**
 * Builder for specifying the target of bulk mappings.
 */
class BulkTargetPathBuilder(
    private val parent: TargetBuilder,
    private val sourcePattern: String,
    private val targetPath: String,
    private val config: BulkMappingConfig
) {
    /**
     * Finalize the bulk mapping rule.
     */
    fun end(): TargetBuilder {
        val rule: MappingRule = createMappingRule()
        return parent.addRule(rule)
    }

    private fun createMappingRule(): MappingRule = when {
        config.flatten -> FlattenMappingRule(
            sourcePattern = sourcePattern,
            targetPath = targetPath,
            exclusions = config.exclusions,
            inclusions = config.inclusions,
            transformation = config.transformation,
            prefix = config.flattenPrefix
        )
        else -> BulkMappingRule(
            sourcePattern = sourcePattern,
            targetPath = targetPath,
            exclusions = config.exclusions,
            inclusions = config.inclusions,
            transformation = config.transformation,
            preserveStructure = config.preserveStructure
        )
    }
}
